File: python/trees/trie.py
"""
Trie (Prefix Tree)
------------------
Insert:  O(m) where m is word length
Search:  O(m)
Prefix:  O(m)

Great for autocomplete, spell checking, prefix-based lookups.
Each node holds a dict of children keyed by character.
"""


class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
        node.is_end = True

    def search(self, word):
        """Returns True if the exact word exists."""
        node = self._find_node(word)
        return node is not None and node.is_end

    def starts_with(self, prefix):
        """Returns True if any word starts with this prefix."""
        return self._find_node(prefix) is not None

    def get_words_with_prefix(self, prefix):
        """Returns all words that start with the given prefix."""
        node = self._find_node(prefix)
        if not node:
            return []

        results = []
        self._collect_words(node, prefix, results)
        return results

    def _find_node(self, prefix):
        node = self.root
        for ch in prefix:
            if ch not in node.children:
                return None
            node = node.children[ch]
        return node

    def _collect_words(self, node, current, results):
        if node.is_end:
            results.append(current)
        for ch, child in node.children.items():
            self._collect_words(child, current + ch, results)

File: javascript/trees/Trie.js
/**
 * Trie (Prefix Tree)
 *
 * Insert/Search/StartsWith: O(m) where m = word length
 */

class TrieNode {
  constructor() {
    this.children = {};
    this.isEnd = false;
  }
}

class Trie {
  constructor() {
    this.root = new TrieNode();
  }

  insert(word) {
    let node = this.root;
    for (const ch of word) {
      if (!node.children[ch]) {
        node.children[ch] = new TrieNode();
      }
      node = node.children[ch];
    }
    node.isEnd = true;
  }

  search(word) {
    const node = this._findNode(word);
    return node !== null && node.isEnd;
  }

  startsWith(prefix) {
    return this._findNode(prefix) !== null;
  }

  getWordsWithPrefix(prefix) {
    const node = this._findNode(prefix);
    if (!node) return [];

    const results = [];
    this._collect(node, prefix, results);
    return results;
  }

  _findNode(prefix) {
    let node = this.root;
    for (const ch of prefix) {
      if (!node.children[ch]) return null;
      node = node.children[ch];
    }
    return node;
  }

  _collect(node, current, results) {
    if (node.isEnd) results.push(current);
    for (const [ch, child] of Object.entries(node.children)) {
      this._collect(child, current + ch, results);
    }
  }
}

module.exports = Trie;
